controllers/osccluster_publicip_controller_unit_test.go

```diff
@@ -0,0 +1,700 @@
+						IpRange: "10.0.0.0/16",
+					},
+					NatService: infrastructurev1beta1.OscNatService{
+						Name:         "test-natservice",
+						PublicIpName: "test-publicip",
+						SubnetName:   "test-subnet",
+					},
+					PublicIps: []*infrastructurev1beta1.OscPublicIp{
+						{
+							Name: "test-publicip-first",
+						},
+						{
+							Name: "test-publicip-first",
+						},
+					},
+				},
+			},
+			expCheckPublicIpOscDuplicateNameErr: fmt.Errorf("test-publicip-first already exist"),
+		},
+	}
+	for _, pitc := range publicIpTestCases {
+		t.Run(pitc.name, func(t *testing.T) {
+			clusterScope := Setup(t, pitc.name, pitc.spec)
+			duplicateResourcePublicIpErr := checkPublicIpOscDuplicateName(clusterScope)
+			if duplicateResourcePublicIpErr != nil {
+				assert.Equal(t, pitc.expCheckPublicIpOscDuplicateNameErr, duplicateResourcePublicIpErr, "checkPublicOscDuplicateName() should return the same error")
+			} else {
+				assert.Nil(t, pitc.expCheckPublicIpOscDuplicateNameErr)
+			}
+		})
+	}
+
+}
+
+// TestReconcilePublicIpGet has several tests to cover the code of the function reconcilePublicIp
+func TestReconcilePublicIpGet(t *testing.T) {
+	publicIpTestCases := []struct {
+		name                    string
+		spec                    infrastructurev1beta1.OscClusterSpec
+		expPublicIpFound        bool
+		expValidatePublicIpsErr error
+		expReconcilePublicIpErr error
+	}{
+		{
+			name:                    "check publicIp exist (second time reconcile loop)",
+			spec:                    defaultPublicIpReconcile,
+			expPublicIpFound:        true,
+			expValidatePublicIpsErr: nil,
+			expReconcilePublicIpErr: nil,
+		},
+		{
+			name:                    "check two publicIp exist (second time reconcile loop)",
+			spec:                    defaultMultiPublicIpReconcile,
+			expPublicIpFound:        true,
+			expValidatePublicIpsErr: nil,
+			expReconcilePublicIpErr: nil,
+		},
+		{
+			name:                    "failed to validate publicIp",
+			spec:                    defaultPublicIpInitialize,
+			expPublicIpFound:        false,
+			expValidatePublicIpsErr: fmt.Errorf("ValidatePublicIp generic error"),
+			expReconcilePublicIpErr: fmt.Errorf("ValidatePublicIp generic error"),
+		},
+	}
+	for _, pitc := range publicIpTestCases {
+		t.Run(pitc.name, func(t *testing.T) {
+			clusterScope, ctx, mockOscPublicIpInterface := SetupWithPublicIpMock(t, pitc.name, pitc.spec)
+			publicIpsSpec := pitc.spec.Network.PublicIps
+			var publicIpIds []string
+			var publicIpId string
+			publicIpRef := clusterScope.GetPublicIpRef()
+			publicIpRef.ResourceMap = make(map[string]string)
+			for _, publicIpSpec := range publicIpsSpec {
+				publicIpName := publicIpSpec.Name + "-uid"
+				publicIpId = "eipalloc-" + publicIpName
+				publicIpIds = append(publicIpIds, publicIpId)
+				if pitc.expPublicIpFound {
+					publicIpRef.ResourceMap[publicIpName] = publicIpId
+				}
+
+			}
+			if pitc.expValidatePublicIpsErr != nil {
+				publicIpIds = []string{""}
+			}
```

Comment:
Are you sure about that ? If we expect an error, we expect to get nothing, no ?

Reply:
Yes
If i remove it will:
```
        Got: [] ([]string)
        Want: is equal to [eipalloc-test-publicip-uid] ([]string)
```
because i do before :
```
			for _, publicIpSpec := range publicIpsSpec {
				publicIpName := publicIpSpec.Name + "-uid"
				publicIpId = "eipalloc-" + publicIpName
				publicIpIds = append(publicIpIds, publicIpId)
				if pitc.expPublicIpFound {
					publicIpRef.ResourceMap[publicIpName] = publicIpId
				}

			}
```
which is need for ```"check publicIp exist (second time reconcile loop)``` and  ```check two publicIp exist (second time reconcile loop)```
I can keep like this or i can put ```failed to validate publicIp``` in a separate function if you prefer ?

Reply:
Ok understood 👍